kong/tools/utils.lua

```diff
@@ -115,11 +120,51 @@ do
   end
 end
 
+local get_rand_bytes
+
 do
+  local ngx_log = ngx.log
+  local WARN    = ngx.WARN
+
+  local system_constants = require "lua_system_constants"
+  local O_RDONLY = system_constants.O_RDONLY()
   local bytes_buf_t = ffi.typeof "char[?]"
 
-  function _M.get_rand_bytes(n_bytes)
+  -- try to get n_bytes of CSPRNG data, first via /dev/urandom,
+  -- and then falling back to OpenSSL if necessary
+  get_rand_bytes = function(n_bytes, urandom)
     local buf = ffi_new(bytes_buf_t, n_bytes)
+    local fallback = false
+
+    -- only read from urandom if we were explicitly asked
+    if urandom then
+      local fd = ffi.C.open("/dev/urandom", O_RDONLY)
+      if fd < 0 then
+        ngx_log(WARN, "Error opening random fd: ",
+                      ffi_str(ffi.C.strerror(ffi.errno())))
+
+        -- fall back to openssl
+        fallback = true
+      end
+
+      local res = ffi.C.read(fd, buf, n_bytes)
```

Comment:
If we fail opening `/dev/urandom`, we still try to read from it, and later on, to close the file descriptor? Wouldn't that leave use with 3 WARN logs instead of one, potentially?